libs/ledger-live-common/src/platform/logic.ts

```diff
@@ -1,4 +1,281 @@
+    canEditFees,
+    liveTx,
+    hasFeesProvided,
+  });
+}
+
+export function broadcastTransactionLogic(
+  { manifest, accounts, tracking }: WebPlatformContext,
+  accountId: string,
+  signedTransaction: RawPlatformSignedTransaction,
+  uiNavigation: (
+    account: AccountLike,
+    parentAccount: Account | null,
+    signedOperation: SignedOperation
+  ) => Promise<string>
+): Promise<string> {
+  if (!signedTransaction) {
+    tracking.platformBroadcastFail(manifest);
+    return Promise.reject(new Error("Transaction required"));
+  }
+
+  const account = accounts.find((account) => account.id === accountId);
+  if (!account) {
+    tracking.platformBroadcastFail(manifest);
+    return Promise.reject(new Error("Account required"));
+  }
+
+  const parentAccount = getParentAccount(account, accounts);
+
+  const signedOperation = deserializePlatformSignedTransaction(
+    signedTransaction,
+    accountId
+  );
+
+  return uiNavigation(account, parentAccount, signedOperation);
+}
+
+export type CompleteExchangeRequest = {
+  provider: string;
+  fromAccountId: string;
+  toAccountId: string;
+  transaction: RawPlatformTransaction;
+  binaryPayload: string;
+  signature: string;
+  feesStrategy: string;
+  exchangeType: number;
+};
+export type CompleteExchangeUiRequest = {
+  provider: string;
+  exchange: {
+    fromAccount: AccountLike;
+    fromParentAccount: Account | null;
+    toAccount?: AccountLike;
+    toParentAccount: Account | null;
+  };
+  transaction: TransactionCommon;
+  binaryPayload: string;
+  signature: string;
+  feesStrategy: string;
+  exchangeType: number;
+};
+export function completeExchangeLogic(
+  { manifest, accounts, tracking }: WebPlatformContext,
+  {
+    provider,
+    fromAccountId,
+    toAccountId,
+    transaction,
+    binaryPayload,
+    signature,
+    feesStrategy,
+    exchangeType,
+  }: CompleteExchangeRequest,
+  uiNavigation: (request: CompleteExchangeUiRequest) => Promise<Operation>
+): Promise<Operation> {
+  tracking.platformCompleteExchangeRequested(manifest);
+
+  // Nb get a hold of the actual accounts, and parent accounts
+  const fromAccount = accounts.find((a) => a.id === fromAccountId);
+
+  const toAccount = accounts.find((a) => a.id === toAccountId);
+
+  if (!fromAccount) {
+    return Promise.reject();
+  }
+
+  if (exchangeType === 0x00 && !toAccount) {
+    // if we do a swap, a destination account must be provided
+    return Promise.reject();
+  }
+
+  const fromParentAccount = getParentAccount(fromAccount, accounts);
+  const toParentAccount = toAccount
+    ? getParentAccount(toAccount, accounts)
+    : null;
+  const exchange = {
+    fromAccount,
+    fromParentAccount,
+    toAccount,
+    toParentAccount,
+  };
+
+  const accountBridge = getAccountBridge(fromAccount, fromParentAccount);
+  const mainFromAccount = getMainAccount(fromAccount, fromParentAccount);
+
+  // TODO-STP: fix this
+  // transaction.family = mainFromAccount.currency.family;
```

Comment:
I need some help to decide what to do with this type error.

Reply:
To be fair, I don't really understand the use / need of this line of code.

It sets the transaction family to the main account currency family. A priori it could be used if the `fromAccountId` provided to the SDK method `completeExchange` does not correspond to the family used by the `transaction` argument of the same method.

For example, I call `completeExchange` with a `fromAccountId` corresponding to a Bitcoin account, but pass an Ethereum transaction for the `transaction` parameter.

But the issue is, this line of code won't help since the transaction would still be an Ethereum transaction (even though the `family` is now `bitcoin`, all the parameters of this transaction are still Ethereum related). So when we call the following code a couple lines bellow, we are actually creating a Bitcoin transaction:
```ts
// `processedTransaction` is a Bitcoin transaction
let processedTransaction = accountBridge.createTransaction(mainFromAccount);
```


But we are then updating this transaction just bellow by calling:
```ts
  processedTransaction = accountBridge.updateTransaction(
// Bitcoin transaction
    processedTransaction,
// Ethereum transaction but with a `bitcoin` value for the `family` field
    platformTransaction
  );
```

This is highly unstable and if it works as of today, it's probably out of sheer luck.

What we could do instead is returning an error in the case there is a mismatch between the account family and the transaction one, like:
```ts
// We could also simply test against `mainFromAccount.family` instead of `currency.family`
  if (transaction.family !== mainFromAccount.currency.family) {
    return Promise.reject("Error");
  }
```